src/math-window.c

```diff
@@ -202,7 +202,27 @@ key_press_cb(MathWindow *window, GdkEventKey *event)
     gboolean result;
     g_signal_emit_by_name(window->priv->display, "key-press-event", event, &result);
 
-    if (math_buttons_get_mode (window->priv->buttons) == PROGRAMMING && (event->state & GDK_CONTROL_MASK) == GDK_CONTROL_MASK) {
+    /* Keyboard navigation for history */
+    if ((event->state & GDK_MOD1_MASK) == GDK_MOD1_MASK && (event->keyval == GDK_KEY_Up || event->keyval == GDK_KEY_Down))
+    {
+        switch (event->keyval)
+        {
+            case GDK_KEY_Up:
+                math_history_set_current(window->priv->history, -1);
+                break;
+            case GDK_KEY_Down:
+                math_history_set_current(window->priv->history, 1);
+                break;
+        }
+
+        MathHistoryEntry *entry = math_history_get_entry_at(window->priv->history, math_history_get_current(window->priv->history));
+        if (entry)
+        {
+            math_equation_set(window->priv->equation, gtk_widget_get_tooltip_text(math_history_entry_get_equation_label(entry)));
```

Comment:
There is a memory leak since you should free the returned value of `gtk_widget_get_tooltip_text`. On the other hand if you use the method `math_history_entry_get_equation_label` to get a string, why do you define a getter for a label? I think it would be better to get directly the string.